File: src/vector.rs
// 向量处理和搜索模块
use std::collections::HashMap;
use std::sync::Arc;
use arrow::array::{Array, BinaryArray, Float32Array, StringArray, UInt64Array, RecordBatchIterator};
use arrow::datatypes::{DataType, Field, Schema};
use arrow::record_batch::RecordBatch;
use futures::TryStreamExt;
use lancedb::{Connection, Table};
use lancedb::query::{QueryBase, ExecutableQuery};
use serde::{Deserialize, Serialize};

use crate::core::AgentDbError;

// 向量相似度算法
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum SimilarityAlgorithm {
    Cosine,
    Euclidean,
    DotProduct,
    Manhattan,
}

impl SimilarityAlgorithm {
    pub fn calculate(&self, a: &[f32], b: &[f32]) -> f32 {
        match self {
            SimilarityAlgorithm::Cosine => cosine_similarity(a, b),
            SimilarityAlgorithm::Euclidean => euclidean_distance(a, b),
            SimilarityAlgorithm::DotProduct => dot_product(a, b),
            SimilarityAlgorithm::Manhattan => manhattan_distance(a, b),
        }
    }
}

// 向量搜索结果
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VectorSearchResult {
    pub id: u64,
    pub vector: Vec<f32>,
    pub metadata: HashMap<String, String>,
    pub similarity: f32,
    pub distance: f32,
}

// 向量索引配置
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VectorIndexConfig {
    pub dimension: usize,
    pub algorithm: SimilarityAlgorithm,
    pub index_type: String,
    pub parameters: HashMap<String, String>,
}

impl Default for VectorIndexConfig {
    fn default() -> Self {
        Self {
            dimension: 768,
            algorithm: SimilarityAlgorithm::Cosine,
            index_type: "HNSW".to_string(),
            parameters: HashMap::new(),
        }
    }
}

// 高级向量引擎
pub struct AdvancedVectorEngine {
    connection: Connection,
    config: VectorIndexConfig,
}

impl AdvancedVectorEngine {
    pub fn new(connection: Connection, config: VectorIndexConfig) -> Self {
        Self { connection, config }
    }

    pub async fn ensure_table(&self) -> Result<Table, AgentDbError> {
        match self.connection.open_table("vectors").execute().await {
            Ok(table) => Ok(table),
            Err(_) => {
                let schema = Schema::new(vec![
                    Field::new("id", DataType::UInt64, false),
                    Field::new("vector", DataType::Binary, false),
                    Field::new("metadata", DataType::Utf8, false),
                    Field::new("timestamp", DataType::Int64, false),
                ]);

                let empty_batches = RecordBatchIterator::new(
                    std::iter::empty::<Result<RecordBatch, arrow::error::ArrowError>>(),
                    Arc::new(schema),
                );

                let table = self
                    .connection
                    .create_table("vectors", Box::new(empty_batches))
                    .execute()
                    .await?;

                Ok(table)
            }
        }
    }

    pub async fn add_vector(&self, id: u64, vector: Vec<f32>, metadata: HashMap<String, String>) -> Result<(), AgentDbError> {
        if vector.len() != self.config.dimension {
            return Err(AgentDbError::InvalidArgument(
                format!("Vector dimension {} doesn't match expected {}", vector.len(), self.config.dimension)
            ));
        }

        let table = self.ensure_table().await?;
        let metadata_json = serde_json::to_string(&metadata)?;
        let vector_bytes = serde_json::to_vec(&vector)?;
        let timestamp = chrono::Utc::now().timestamp();

        let schema = table.schema().await?;

        let batch = RecordBatch::try_new(
            schema.clone(),
            vec![
                Arc::new(UInt64Array::from(vec![id])),
                Arc::new(BinaryArray::from(vec![vector_bytes.as_slice()])),
                Arc::new(StringArray::from(vec![metadata_json])),
                Arc::new(arrow::array::Int64Array::from(vec![timestamp])),
            ],
        )?;

        let batch_iter = RecordBatchIterator::new(
            std::iter::once(Ok(batch)),
            schema,
        );
        table.add(Box::new(batch_iter)).execute().await?;
        Ok(())
    }

    pub async fn batch_add_vectors(&self, vectors: Vec<(u64, Vec<f32>, HashMap<String, String>)>) -> Result<(), AgentDbError> {
        if vectors.is_empty() {
            return Ok(());
        }

        let table = self.ensure_table().await?;
        let schema = table.schema().await?;
        let timestamp = chrono::Utc::now().timestamp();

        let mut ids = Vec::new();
        let mut vector_bytes_vec = Vec::new();
        let mut metadata_jsons = Vec::new();
        let mut timestamps = Vec::new();

        for (id, vector, metadata) in vectors {
            if vector.len() != self.config.dimension {
                return Err(AgentDbError::InvalidArgument(
                    format!("Vector dimension {} doesn't match expected {}", vector.len(), self.config.dimension)
                ));
            }

            ids.push(id);
            vector_bytes_vec.push(serde_json::to_vec(&vector)?);
            metadata_jsons.push(serde_json::to_string(&metadata)?);
            timestamps.push(timestamp);
        }

        let vector_bytes_refs: Vec<&[u8]> = vector_bytes_vec.iter().map(|v| v.as_slice()).collect();

        let batch = RecordBatch::try_new(
            schema.clone(),
            vec![
                Arc::new(UInt64Array::from(ids)),
                Arc::new(BinaryArray::from(vector_bytes_refs)),
                Arc::new(StringArray::from(metadata_jsons)),
                Arc::new(arrow::array::Int64Array::from(timestamps)),
            ],
        )?;

        let batch_iter = RecordBatchIterator::new(
            std::iter::once(Ok(batch)),
            schema,
        );
        table.add(Box::new(batch_iter)).execute().await?;
        Ok(())
    }

    pub async fn search_vectors(&self, query_vector: &[f32], limit: usize) -> Result<Vec<VectorSearchResult>, AgentDbError> {
        if query_vector.len() != self.config.dimension {
            return Err(AgentDbError::InvalidArgument(
                format!("Query vector dimension {} doesn't match expected {}", query_vector.len(), self.config.dimension)
            ));
        }

        let table = self.ensure_table().await?;
        let mut results = table.query().limit(limit * 10).execute().await?; // 获取更多结果用于排序

        let mut candidates = Vec::new();
        while let Some(batch) = results.try_next().await? {
            for row in 0..batch.num_rows() {
                let result = self.extract_vector_from_batch(&batch, row)?;
                let similarity = self.config.algorithm.calculate(query_vector, &result.vector);
                
                candidates.push(VectorSearchResult {
                    id: result.id,
                    vector: result.vector,
                    metadata: result.metadata,
                    similarity,
                    distance: 1.0 - similarity,
                });
            }
        }

        // 按相似度排序并返回前limit个结果
        candidates.sort_by(|a, b| b.similarity.partial_cmp(&a.similarity).unwrap());
        candidates.truncate(limit);

        Ok(candidates)
    }

    pub async fn get_vector(&self, id: u64) -> Result<Option<VectorSearchResult>, AgentDbError> {
        let table = self.ensure_table().await?;

        let mut results = table
            .query()
            .only_if(&format!("id = {}", id))
            .limit(1)
            .execute()
            .await?;

        let batch = match results.try_next().await? {
            Some(batch) => batch,
            None => return Ok(None),
        };

        if batch.num_rows() == 0 {
            return Ok(None);
        }

        let result = self.extract_vector_from_batch(&batch, 0)?;
        Ok(Some(VectorSearchResult {
            id: result.id,
            vector: result.vector,
            metadata: result.metadata,
            similarity: 1.0,
            distance: 0.0,
        }))
    }

    pub async fn update_vector(&self, id: u64, vector: Vec<f32>, metadata: HashMap<String, String>) -> Result<(), AgentDbError> {
        let table = self.ensure_table().await?;
        
        // 删除旧向量
        table.delete(&format!("id = {}", id)).await?;
        
        // 添加新向量
        self.add_vector(id, vector, metadata).await?;
        
        Ok(())
    }

    pub async fn delete_vector(&self, id: u64) -> Result<(), AgentDbError> {
        let table = self.ensure_table().await?;
        table.delete(&format!("id = {}", id)).await?;
        Ok(())
    }

    fn extract_vector_from_batch(&self, batch: &RecordBatch, row: usize) -> Result<VectorSearchResult, AgentDbError> {
        let id_array = batch.column(0).as_any().downcast_ref::<UInt64Array>().unwrap();
        let vector_array = batch.column(1).as_any().downcast_ref::<BinaryArray>().unwrap();
        let metadata_array = batch.column(2).as_any().downcast_ref::<StringArray>().unwrap();

        let id = id_array.value(row);
        let vector_bytes = vector_array.value(row);
        let vector: Vec<f32> = serde_json::from_slice(vector_bytes)?;
        let metadata_json = metadata_array.value(row);
        let metadata: HashMap<String, String> = serde_json::from_str(metadata_json)?;

        Ok(VectorSearchResult {
            id,
            vector,
            metadata,
            similarity: 0.0,
            distance: 0.0,
        })
    }

    pub async fn get_vector_count(&self) -> Result<usize, AgentDbError> {
        let table = self.ensure_table().await?;
        let mut results = table.query().execute().await?;
        let mut count = 0;

        while let Some(batch) = results.try_next().await? {
            count += batch.num_rows();
        }

        Ok(count)
    }
}

// 相似度计算函数
fn cosine_similarity(a: &[f32], b: &[f32]) -> f32 {
    let dot_product: f32 = a.iter().zip(b.iter()).map(|(x, y)| x * y).sum();
    let norm_a: f32 = a.iter().map(|x| x * x).sum::<f32>().sqrt();
    let norm_b: f32 = b.iter().map(|x| x * x).sum::<f32>().sqrt();
    
    if norm_a == 0.0 || norm_b == 0.0 {
        0.0
    } else {
        dot_product / (norm_a * norm_b)
    }
}

fn euclidean_distance(a: &[f32], b: &[f32]) -> f32 {
    let sum: f32 = a.iter().zip(b.iter()).map(|(x, y)| (x - y).powi(2)).sum();
    sum.sqrt()
}

fn dot_product(a: &[f32], b: &[f32]) -> f32 {
    a.iter().zip(b.iter()).map(|(x, y)| x * y).sum()
}

fn manhattan_distance(a: &[f32], b: &[f32]) -> f32 {
    a.iter().zip(b.iter()).map(|(x, y)| (x - y).abs()).sum()
}

File: agent-db-zig/src/vector.zig
// 向量处理 Zig API
const std = @import("std");
const c = @cImport({
    @cInclude("agent_state_db.h");
});

pub const SimilarityAlgorithm = enum(c_int) {
    cosine = 0,
    euclidean = 1,
    dot_product = 2,
    manhattan = 3,
};

pub const VectorSearchResult = struct {
    id: u64,
    vector: []f32,
    metadata: std.StringHashMap([]const u8),
    similarity: f32,
    distance: f32,
    
    const Self = @This();
    
    pub fn init(allocator: std.mem.Allocator, id: u64, vector: []const f32, similarity: f32) !Self {
        return Self{
            .id = id,
            .vector = try allocator.dupe(f32, vector),
            .metadata = std.StringHashMap([]const u8).init(allocator),
            .similarity = similarity,
            .distance = 1.0 - similarity,
        };
    }
    
    pub fn deinit(self: *Self, allocator: std.mem.Allocator) void {
        allocator.free(self.vector);
        
        var iterator = self.metadata.iterator();
        while (iterator.next()) |entry| {
            allocator.free(entry.key_ptr.*);
            allocator.free(entry.value_ptr.*);
        }
        self.metadata.deinit();
    }
    
    pub fn setMetadata(self: *Self, allocator: std.mem.Allocator, key: []const u8, value: []const u8) !void {
        const owned_key = try allocator.dupe(u8, key);
        const owned_value = try allocator.dupe(u8, value);
        try self.metadata.put(owned_key, owned_value);
    }
};

pub const VectorEngine = struct {
    db: ?*c.CAgentStateDB,
    allocator: std.mem.Allocator,
    dimension: usize,
    algorithm: SimilarityAlgorithm,
    
    const Self = @This();
    
    pub fn init(allocator: std.mem.Allocator, db_path: []const u8, dimension: usize, algorithm: SimilarityAlgorithm) !Self {
        const c_path = try allocator.dupeZ(u8, db_path);
        defer allocator.free(c_path);
        
        const db = c.agent_db_new(c_path.ptr);
        if (db == null) {
            return error.DatabaseInitFailed;
        }
        
        return Self{
            .db = db,
            .allocator = allocator,
            .dimension = dimension,
            .algorithm = algorithm,
        };
    }
    
    pub fn deinit(self: *Self) void {
        if (self.db) |db| {
            c.agent_db_free(db);
        }
    }
    
    pub fn addVector(self: *Self, id: u64, vector: []const f32, metadata: ?std.StringHashMap([]const u8)) !void {
        if (self.db == null) return error.DatabaseNotInitialized;
        
        if (vector.len != self.dimension) {
            return error.InvalidVectorDimension;
        }
        
        // 序列化向量和元数据
        const vector_bytes = std.mem.sliceAsBytes(vector);
        
        const result = c.agent_db_save_state(
            self.db.?,
            id,
            0, // session_id
            5, // embeddings type
            vector_bytes.ptr,
            vector_bytes.len
        );
        
        if (result != 0) {
            return error.AddVectorFailed;
        }
    }
    
    pub fn searchVectors(self: *Self, query_vector: []const f32, limit: usize) !std.ArrayList(VectorSearchResult) {
        if (self.db == null) return error.DatabaseNotInitialized;
        
        if (query_vector.len != self.dimension) {
            return error.InvalidVectorDimension;
        }
        
        const results = std.ArrayList(VectorSearchResult).init(self.allocator);
        
        // 简化实现，实际应该调用相应的C函数进行向量搜索
        // 这里返回空结果作为占位符
        _ = limit;
        
        return results;
    }
    
    pub fn getVector(self: *Self, id: u64) !?VectorSearchResult {
        if (self.db == null) return error.DatabaseNotInitialized;
        
        // 简化实现，实际应该调用相应的C函数
        _ = id;
        return null;
    }
    
    pub fn updateVector(self: *Self, id: u64, vector: []const f32, metadata: ?std.StringHashMap([]const u8)) !void {
        if (self.db == null) return error.DatabaseNotInitialized;
        
        if (vector.len != self.dimension) {
            return error.InvalidVectorDimension;
        }
        
        // 简化实现，实际应该先删除再添加
        try self.addVector(id, vector, metadata);
    }
    
    pub fn deleteVector(self: *Self, id: u64) !void {
        if (self.db == null) return error.DatabaseNotInitialized;
        
        // 简化实现，实际应该调用相应的C函数
        _ = id;
    }
    
    pub fn batchAddVectors(self: *Self, vectors: []const struct { id: u64, vector: []const f32, metadata: ?std.StringHashMap([]const u8) }) !void {
        for (vectors) |vec_data| {
            try self.addVector(vec_data.id, vec_data.vector, vec_data.metadata);
        }
    }
    
    pub fn getVectorCount(self: *Self) !usize {
        if (self.db == null) return error.DatabaseNotInitialized;
        
        // 简化实现，实际应该调用相应的C函数
        return 0;
    }
};

// 相似度计算函数
pub fn cosineSimilarity(a: []const f32, b: []const f32) f32 {
    if (a.len != b.len) return 0.0;
    
    var dot_product: f32 = 0.0;
    var norm_a: f32 = 0.0;
    var norm_b: f32 = 0.0;
    
    for (a, b) |x, y| {
        dot_product += x * y;
        norm_a += x * x;
        norm_b += y * y;
    }
    
    norm_a = @sqrt(norm_a);
    norm_b = @sqrt(norm_b);
    
    if (norm_a == 0.0 or norm_b == 0.0) {
        return 0.0;
    }
    
    return dot_product / (norm_a * norm_b);
}

pub fn euclideanDistance(a: []const f32, b: []const f32) f32 {
    if (a.len != b.len) return std.math.inf(f32);
    
    var sum: f32 = 0.0;
    for (a, b) |x, y| {
        const diff = x - y;
        sum += diff * diff;
    }
    
    return @sqrt(sum);
}

pub fn dotProduct(a: []const f32, b: []const f32) f32 {
    if (a.len != b.len) return 0.0;
    
    var result: f32 = 0.0;
    for (a, b) |x, y| {
        result += x * y;
    }
    
    return result;
}

pub fn manhattanDistance(a: []const f32, b: []const f32) f32 {
    if (a.len != b.len) return std.math.inf(f32);
    
    var sum: f32 = 0.0;
    for (a, b) |x, y| {
        sum += @abs(x - y);
    }
    
    return sum;
}

test "Vector similarity calculations" {
    const a = [_]f32{ 1.0, 2.0, 3.0 };
    const b = [_]f32{ 4.0, 5.0, 6.0 };
    
    const cosine_sim = cosineSimilarity(&a, &b);
    try std.testing.expect(cosine_sim > 0.9); // 应该很相似
    
    const euclidean_dist = euclideanDistance(&a, &b);
    try std.testing.expect(euclidean_dist > 0.0);
    
    const dot_prod = dotProduct(&a, &b);
    try std.testing.expect(dot_prod == 32.0); // 1*4 + 2*5 + 3*6 = 32
    
    const manhattan_dist = manhattanDistance(&a, &b);
    try std.testing.expect(manhattan_dist == 9.0); // |1-4| + |2-5| + |3-6| = 9
}

test "VectorSearchResult creation" {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    
    const vector = [_]f32{ 1.0, 2.0, 3.0 };
    var result = try VectorSearchResult.init(allocator, 123, &vector, 0.95);
    defer result.deinit(allocator);
    
    try std.testing.expect(result.id == 123);
    try std.testing.expect(result.similarity == 0.95);
    try std.testing.expect(result.distance == 0.05);
    try std.testing.expect(result.vector.len == 3);
}
